runtime/src/main/java/org/corfudb/runtime/LogReplicationRoutingQueueListener.java

```diff
@@ -0,0 +1,357 @@
+
+    private final CorfuStore corfuStore;
+    private final String namespace;
+
+    private CorfuStoreMetadata.Timestamp lastProcessedEntryTs;
+
+    /**
+     * Special LogReplication listener which a client creates to receive ordered updates for replicated data.
+     * @param corfuStore Corfu Store used on the client
+     * @param namespace Namespace of the client's tables
+     */
+    public LogReplicationRoutingQueueListener(CorfuStore corfuStore, @Nonnull String namespace) {
+        this.corfuStore = corfuStore;
+        this.namespace = namespace;
+    }
+
+    /**
+     * This is an internal method of this abstract listener and not exposed to clients.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public final void onNext(CorfuStreamEntries results) {
+
+        // If this update came before the client's full sync timestamp, ignore it.
+        if (results.getTimestamp().getSequence() <= clientFullSyncTimestamp.get()) {
+            return;
+        }
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE_NAME)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                    "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (clientFullSyncPending.get()) {
+            // If the listener started when snapshot sync was ongoing, ignore all data updates until it ends.  When
+            // it ends, the client will perform a full sync and build a consistent state containing these updates.
+            return;
+        }
+
+        boolean callbackResult = false;
+        List<Queue.RoutingTableEntryMsg> entries =
+                results.getEntries().entrySet().stream().map(CSEntry ->
+                        (Queue.RoutingTableEntryMsg)CSEntry.getValue().get(0).getPayload()).collect(Collectors.toList());
+        if (snapshotSyncInProgress.get()) {
+            // For routing queue listener, we have only 1 entry inside corfuStreamEntries.
+            callbackResult = processUpdatesInSnapshotSync(entries);
+        } else {
+            callbackResult = processUpdatesInLogEntrySync(entries);
+        }
+        if (callbackResult) {
+            // When the client processed the entry successfully, delete the entry from the routing queue.
+            deleteEntryFromQueue(results);
+        }
+        this.lastProcessedEntryTs = results.getTimestamp();
+    }
+
+    private void deleteEntryFromQueue(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+        List<CorfuStreamEntry> routingQueueTableEntries =
+                entries.entrySet().stream().filter(
```

Comment:
why do we require filtering?

Reply:
no need to. Seems we'll have only one queue present at this time. Removing the filtering